src/lib.rs

```diff
@@ -4169,11 +4211,11 @@ struct PrintThread {
 }
 
 impl PrintThread {
-    fn new() -> Result<Self, Error> {
+    fn new(cargo_warnings: bool) -> Result<Self, Error> {
         let (pipe_reader, pipe_writer) = os_pipe::pipe()?;
```

Comment:
I recommend to skip spawaning the thread to reduce cost of compilation, if warning is disabled.

You could also skip creating pipes, but then you need to change code using `pipe_writer*`

Reply:
I saw this too but cautiously decided against it because `PrintThread` also does line-by-line locking of stdout. I'm worried that removing this will cause garbled output when using the `parallel` feature.

Reply:
PrintThread only lock stdout when it receives a line to print to stdout, tomprevent interleaving.

Reply:
Oh but if we take the approach of fully disabling output instead of just removing the instruction prefix, we can just pass around an `Option<PrintThread>` and close stderr if it's `None`.